src/configure-balena.sh

```diff
@@ -0,0 +1,332 @@
+SENTRY_ENVVARS[IMG_DSN]=6
+SENTRY_ENVVARS[PROXY_DSN]=7
+SENTRY_ENVVARS[UI_DSN]=8
+SENTRY_ENVVARS[VPN_DSN]=9
+
+# API keys
+declare -A API_KEYS
+API_KEYS[API_SERVICE_API_KEY]="$(openssl rand -hex 16)"
+API_KEYS[AUTH_RESINOS_REGISTRY_CODE]="$(openssl rand -hex 16)"
+API_KEYS[BUILDER_SERVICE_API_KEY]="$(openssl rand -hex 16)"
+API_KEYS[COOKIE_SESSION_SECRET]="$(openssl rand -hex 16)"
+API_KEYS[DELTA_SERVICE_API_KEY]="$(openssl rand -hex 16)"
+API_KEYS[DIGITIZER_API_KEY]="$(openssl rand -hex 16)"
+API_KEYS[GEOIP_LICENCE_KEY]="$(openssl rand -hex 16)"
+API_KEYS[GEOIP_USER_ID]="$(openssl rand -hex 16)"
+API_KEYS[GIT_API_KEY]="$(openssl rand -hex 16)"
+API_KEYS[IMG_S3_ACCESS_KEY]="$(openssl rand -hex 16)"
+API_KEYS[IMG_S3_SECRET_KEY]="$(openssl rand -hex 16)"
+API_KEYS[JF_OAUTH_APP_SECRET]="$(openssl rand -hex 16)"
+API_KEYS[JSON_WEB_TOKEN_SECRET]="$(openssl rand -hex 16)"
+API_KEYS[MAPS_API_KEY]="$(openssl rand -hex 16)"
+API_KEYS[MIXPANEL_TOKEN]="$(openssl rand -hex 16)"
+API_KEYS[MONITOR_OAUTH_COOKIE_SECRET]="$(openssl rand -hex 16)"
+API_KEYS[MONITOR_SECRET_TOKEN]="$(openssl rand -hex 16)"
+API_KEYS[PROXY_SERVICE_API_KEY]="$(openssl rand -hex 16)"
+API_KEYS[REGISTRY2_SECRETKEY]="$(openssl rand -hex 16)"
+API_KEYS[SENTRY_ADMIN_APIKEY]="$(openssl rand -hex 16)"
+API_KEYS[SENTRY_SECRET_KEY]="$(openssl rand -hex 16)"
+API_KEYS[TOKEN_AUTH_BUILDER_TOKEN]="$(openssl rand -hex 16)"
+API_KEYS[VPN_GUEST_API_KEY]="$(openssl rand -hex 16)"
+API_KEYS[VPN_SERVICE_API_KEY]="$(openssl rand -hex 16)"
+API_KEYS[API_VPN_SERVICE_API_KEY]="${API_KEYS[API_SERVICE_API_KEY]}"
+API_KEYS[DELTA_API_KEY]="${API_KEYS[DELTA_SERVICE_API_KEY]}"
+API_KEYS[DELTA_S3_KEY]="${API_KEYS[IMG_S3_ACCESS_KEY]}"
+API_KEYS[DELTA_S3_SECRET]="${API_KEYS[IMG_S3_SECRET_KEY]}"
+API_KEYS[GIT_SERVICE_API_KEY]="${API_KEYS[GIT_API_KEY]}"
+API_KEYS[MDNS_API_TOKEN]="${API_KEYS[PROXY_SERVICE_API_KEY]}"
+API_KEYS[REGISTRY2_TOKEN]="${API_KEYS[TOKEN_AUTH_BUILDER_TOKEN]}"
+API_KEYS[S3_MINIO_ACCESS_KEY]="${API_KEYS[IMG_S3_ACCESS_KEY]}"
+API_KEYS[S3_MINIO_SECRET_KEY]="${API_KEYS[IMG_S3_SECRET_KEY]}"
+
+function upsert_ca_root {
+	if test -n "${ROOT_CA}"; then
+		echo "Installing custom CA bundle..."
+		echo "${ROOT_CA}" | base64 -d > /usr/local/share/ca-certificates/balenaRootCA.crt
+	elif [[ -e "${CERTS}/ca-bundle.pem" ]]; then
+		cat < "${CERTS}/ca-bundle.pem" > /usr/local/share/ca-certificates/balenaRootCA.crt
+		sed -i /^ROOT_CA=/d /etc/docker.env
+		echo "ROOT_CA=$(cat < "${CERTS}/ca-bundle.pem" | openssl base64 -A)" >> /etc/docker.env
+	else
+		echo "Custom CA bundle not found, nothing to do."
+	fi
+
+	if [[ -e /usr/local/share/ca-certificates/balenaRootCA.crt ]]; then
+		grep -q 'NODE_EXTRA_CA_CERTS=/etc/ssl/certs/balenaRootCA.pem' /etc/docker.env \
+		  || echo 'NODE_EXTRA_CA_CERTS=/etc/ssl/certs/balenaRootCA.pem' >> /etc/docker.env
+
+		update-ca-certificates
+	fi
+}
+
+function upsert_api_key {
+	if [[ -f "${CERTS}/private/api.${BALENA_DEVICE_UUID}.${DNS_TLD}.key" ]]; then
```

Comment:
Can you please require `BALENA_DEVICE_UUID`, `DNS_TLD`, etc. to be given as input to the function? This would help with keeping the functions simple/shallow/pure-ish and surface exactly what each one needs, which means we'll be able to spot patterns/commonality and refactor further. You are invoking each from a single place so it hopefully won't get unwieldy.

I'm fine with skipping this for this PR, but I think that we should ultimately get there and this looks like a good opportunity.

Reply:
I've refactored all the functions.